# K-Nearest Neighbors (K-NN) và Naive Bayes


## 1. Thuật toán K-Nearest Neighbors (K-NN)
K-Nearest Neighbors (K-NN) là một thuật toán học có giám sát (supervised learning) dùng cho cả bài toán phân lớp (classification) và hồi quy (regression). Ý tưởng chính của K-NN là: một mẫu cần dự đoán sẽ được gán nhãn dựa trên các mẫu “láng giềng gần nhất” với nó trong không gian đặc trưng.




*   Nguyên lý hoạt động
    1. Cho trước tập dữ liệu huấn luyện gồm các mẫu đã biết nhãn
    2. Khi cần dự đoán cho một mẫu mới, thuật toán:
        - Tính khoảng cách giữa mẫu mới và tất cả các mẫu trong tập huấn luyện (thường dùng khoảng cách Euclid, Manhattan, Minkowski,…).
        - Chọn ra K mẫu gần nhất (K-nearest neighbors).
    3. Với bài toán phân lớp
     - Nhãn của mẫu mới được quyết định theo đa số nhãn trong K láng giềng (majority voting).
    4. Với bài toán hồi quy:
    - Giá trị cần dự đoán thường là trung bình (hoặc trung vị) của các giá trị mục tiêu trong K láng giềng.
    
*   Đặc điểm
- K-NN là thuật toán “lười biếng” (lazy learning): không xây dựng mô hình rõ ràng trong giai đoạn huấn luyện, mà chỉ lưu trữ dữ liệu. Quá trình tính toán chủ yếu diễn ra ở giai đoạn dự đoán.
- Rất trực quan và dễ cài đặt, không yêu cầu giả định mạnh về phân phối dữ liệu.
- Việc lựa chọn K và hàm đo khoảng cách ảnh hưởng lớn đến hiệu quả:
  - K quá nhỏ → mô hình dễ bị nhiễu, overfitting.
  - K quá lớn → mô hình quá “mịn”, dễ underfitting.


## 2. Thuật toán Naive Bayes

Naive Bayes là một họ các thuật toán phân lớp thuộc nhóm học có giám sát, dựa trên **định lý Bayes** với giả định “ngây thơ” rằng các đặc trưng (features) là **độc lập có điều kiện** khi đã biết nhãn lớp.  
Mặc dù giả định này thường không hoàn toàn đúng trong thực tế, Naive Bayes vẫn hoạt động tốt trong nhiều bài toán, đặc biệt là **phân loại văn bản**.

### 2.1. Cơ sở lý thuyết (Định lý Bayes)

Với một mẫu cần phân lớp được mô tả bởi vector đặc trưng:

$$
X = (x_1, x_2, \dots, x_n)
$$

và tập các lớp:

$$
C = \{c_1, c_2, \dots, c_k\}
$$

Ta muốn tìm lớp có **xác suất hậu nghiệm** lớn nhất:

$$
\hat{c} = \arg\max_{c_i \in C} P(c_i \mid X)
$$

Theo **định lý Bayes**:

$$
P(c_i \mid X) = \frac{P(X \mid c_i) \, P(c_i)}{P(X)}
$$

Vì $P(X)$ là hằng số khi so sánh giữa các lớp, ta có:

$$
P(c_i \mid X) \propto P(X \mid c_i) \, P(c_i)
$$

Thuật toán Naive Bayes giả định rằng các thuộc tính $x_j$ **độc lập có điều kiện** với nhau khi biết lớp $c_i$:

$$
P(X \mid c_i) = P(x_1, x_2, \dots, x_n \mid c_i) = \prod_{j=1}^{n} P(x_j \mid c_i)
$$

Do đó, ta có:

$$
\hat{c} = \arg\max_{c_i \in C} \left[ P(c_i) \prod_{j=1}^{n} P(x_j \mid c_i) \right]
$$

Đây là công thức cơ bản dùng trong phân loại bằng Naive Bayes.

### 2.2. Quy trình hoạt động

**Huấn luyện (Training)**  
Từ dữ liệu huấn luyện, ước lượng:

- Xác suất tiên nghiệm cho mỗi lớp:
  $$
  P(c_i) = \frac{\text{số mẫu thuộc lớp } c_i}{\text{tổng số mẫu}}
  $$
- Các xác suất có điều kiện:
  $$
  P(x_j \mid c_i)
  $$
  (tùy biến thể: rời rạc, nhị phân, hay liên tục).

**Dự đoán (Prediction)**  

- Với một mẫu mới $X = (x_1, \dots, x_n)$, tính:
  $$
  P(c_i \mid X) \propto P(c_i)\prod_{j=1}^{n}P(x_j \mid c_i)
  $$
- Chọn lớp có giá trị lớn nhất:
  $$
  \hat{c} = \arg\max_{c_i \in C} \left[ P(c_i) \prod_{j=1}^{n} P(x_j \mid c_i) \right]
  $$

### 2.3. Một số biến thể Naive Bayes phổ biến

- **Multinomial Naive Bayes**
  - Thường dùng cho phân loại văn bản (bag-of-words, TF-IDF).
  - Mô hình hoá số lần xuất hiện của từ trong văn bản.

- **Bernoulli Naive Bayes**
  - Dùng cho dữ liệu nhị phân (một thuộc tính có/không, xuất hiện/không xuất hiện).

- **Gaussian Naive Bayes**
  - Dùng cho thuộc tính liên tục.
  - Giả định mỗi thuộc tính trong từng lớp tuân theo phân phối chuẩn (Gaussian).

# Cài đặt lại k-nearest, naive bayes theo hướng dẫn đính kèm

## Cài đặt lại K-Nearest

In [ ]:
import numpy as np
from collections import Counter

class K_Nearest_Neighbors:
  def __init__(self, data_set, k):
    self.ds = data_set
    self.k = k

  def predict(self, feature_set):
    distances = []
    for group in self.ds:
      for feature in self.ds[group]:

        e_d = np.linalg.norm(np.array(feature) - np.array(feature_set))
        distances.append([e_d, group])

    nearest = sorted(distances)[:self.k]

    votes = [d[1] for d in nearest]

    nearest_group = Counter(votes).most_common(1)[0]
    feature_set_group, self.confidence = nearest_group[0], nearest_group[1] / self.k
    return feature_set_group

  def test(self, test_data):
    correct = 0
    total = 0
    for group in test_data:
      for feature_set in test_data[group]:
        group_prediction = self.predict(feature_set)
        if group_prediction == group:
          correct += 1
        total += 1
    accuracy = correct / total
    print("Accuracy=", accuracy)


import pandas as pd
import random



IRIS_PATH = "https://drive.google.com/uc?export=download&id=14o6FAKLOAvqZiLlmjxld10SMMuQoUJqe"
iris_data = pd.read_csv(IRIS_PATH)


iris_data = iris_data.drop(['Id'], axis=1)


data_set = iris_data.values.tolist()

random.shuffle(data_set)

train_data = {"Iris-setosa": [], "Iris-versicolor": [], "Iris-virginica": []}
test_data = {"Iris-setosa": [], "Iris-versicolor": [], "Iris-virginica": []}
for i in data_set[:120]:
  train_data[i[-1]].append(i[:-1])
for i in data_set[120:]:
  test_data[i[-1]].append(i[:-1])

# MODEL
knn = K_Nearest_Neighbors(train_data, k=7)
knn.test(test_data)

Accuracy= 0.9333333333333333


## Cài đặt lại Naive Bayes

In [ ]:
import numpy as np
import pandas as pd
import random

class Naive_Bayes:
  def __init__(self, data_set):
    self.ds = data_set

    self.ds_means = self.ds.groupby(4).mean(numeric_only=True)
    self.ds_variances = self.ds.groupby(4).var(numeric_only=True)
    self.class_probabilities = self.get_class_probabilities(self.ds)

  def get_class_probabilities(self, data_set):
    class_sizes = data_set.groupby(4).size()
    ds_total = data_set.shape[0]
    probs = {}

    for class_name, size in class_sizes.items():
      probs[class_name] = size / ds_total
    return probs

  def get_probability_density(self, x, mean, variance):
    pd_val = 1 / (np.sqrt(2 * np.pi * variance)) * np.exp((-(x - mean)**2) / (2 * variance))
    return pd_val

  def predict(self, x):
    feature_class_probabilities = {}
    for group, class_prob in self.class_probabilities.items():
      feature_class_probabilities[group] = class_prob
      for i in range(len(x)):

        mean = self.ds_means.loc[group][i]
        var = self.ds_variances.loc[group][i]
        feature_class_probabilities[group] *= self.get_probability_density(x[i], mean, var)

    return max(feature_class_probabilities, key=feature_class_probabilities.get)

  def test(self, test_data):
    correct = 0
    total = 0
    for row in test_data.itertuples():
      feature_set = row[1:5]
      actual_class = row[5]

      group = self.predict(feature_set)

      if group == actual_class:
        correct += 1
      else:
        pass

      total += 1

    accuracy = correct / total
    print("Accuracy=", accuracy)


IRIS_PATH = "https://drive.google.com/uc?export=download&id=14o6FAKLOAvqZiLlmjxld10SMMuQoUJqe"
iris_data = pd.read_csv(IRIS_PATH)

iris_data = iris_data.drop(['Id'], axis=1)

data_set = iris_data.values.tolist()
random.shuffle(data_set)

train_data = pd.DataFrame(data_set[:120])
test_data = pd.DataFrame(data_set[120:])

nb = Naive_Bayes(train_data)
nb.test(test_data)

Accuracy= 1.0


# Các bài tập và nhận xét báo cáo, code thí nghiệm

## THỰC HÀNH VỀ PHÂN LỚP K NEAREST NEIGHBORS


In [ ]:
# Load dataset here"
import pandas as pd

# Convert link Google Drive
IRIS_PATH = "https://drive.google.com/uc?export=download&id=14o6FAKLOAvqZiLlmjxld10SMMuQoUJqe"
LETTER_PATH = "https://drive.google.com/uc?export=download&id=1nSyFxnhYs-epKhPlzTsULtsZGquDeFlG"

iris_data = pd.read_csv(IRIS_PATH)
lc_datga = pd.read_csv(LETTER_PATH)

print("Load data thành công!")
print("Iris_data shape:", iris_data.shape)
print("Letter_recognition shape:", lc_datga.shape)
print(iris_data.head())

Load data thành công!
Iris_data shape: (150, 6)
Letter_recognition shape: (20000, 17)
   Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa


### Câu 1. Toy example: Phân loại loài hoa dựa trên tập dữ liệu Iris. Cài đặt chương trình demo thuật toán K Nearest Neighbors. Tham khảo: https://alphacoder.xyz/k-nearest-neighbors, https://archive.ics.uci.edu/ml/datasets/iris

## Import lib và load dữ liệu

In [ ]:
import numpy as np
import pandas as pd

from collections import Counter
# import uci data
from sklearn.datasets import load_iris
iris_data = load_iris()
df = pd.DataFrame(iris_data.data, columns=iris_data.feature_names)
df['target'] = iris_data.target
df = df.sample(frac=1).reset_index(drop=True)

In [ ]:
df.head()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
0,4.9,3.6,1.4,0.1,0
1,5.7,2.9,4.2,1.3,1
2,6.3,2.7,4.9,1.8,2
3,5.7,4.4,1.5,0.4,0
4,5.6,3.0,4.5,1.5,1


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   target             150 non-null    int64  
dtypes: float64(4), int64(1)
memory usage: 6.0 KB


## Chia dữ liệu

In [ ]:
from sklearn.model_selection import train_test_split

X = df.drop("target", axis=1)
y = df["target"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

## Chuẩn hóa dữ liệu

In [ ]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

## Huấn luyện KNN

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

model = KNeighborsClassifier(n_neighbors=5)  # k = 5 là phổ biến
model.fit(X_train, y_train)

KNeighborsClassifier()

## Đánh giá mô hình

In [ ]:
from sklearn.metrics import accuracy_score

y_pred = model.predict(X_test)
acc = accuracy_score(y_test, y_pred)
print("Accuracy:", acc)

Accuracy: 0.9333333333333333


## Thử nghiệm mô hình trên dữ liệu mẫu

In [ ]:
import numpy as np

sample_data = np.array([[5.1, 3.5, 1.4, 0.2]]) # Ví dụ cho Iris-setosa

# Chuẩn hóa dữ liệu mẫu bằng scaler đã huấn luyện
sample_scaled = scaler.transform(sample_data)

# Dự đoán lớp của dữ liệu mẫu
prediction = model.predict(sample_scaled)

predicted_species = iris_data.target_names[prediction[0]]

print(f"Dữ liệu mẫu: {sample_data[0]}")
print(f"Lớp dự đoán: {prediction[0]} ({predicted_species})")

Dữ liệu mẫu: [5.1 3.5 1.4 0.2]
Lớp dự đoán: 0 (setosa)


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


## Báo cáo ,nhận xét:
Dựa trên kết quả huấn luyện và đánh giá, đây là một số nhận xét về mô hình K-Nearest Neighbors (KNN) mà bạn đã triển khai:

Nhận xét chung về Mô hình K-Nearest Neighbors (KNN)
- 1. Đối với bài toán phân loại loài hoa Iris:

Độ chính xác (Accuracy): 0.9333 (khoảng 93.33%).
Nhận xét: Đây là một kết quả rất tốt cho bài toán phân loại 3 loài hoa Iris. Với một tập dữ liệu tương đối nhỏ và rõ ràng như Iris, KNN hoạt động hiệu quả. Việc chuẩn hóa dữ liệu (StandardScaler) trước khi huấn luyện là một bước quan trọng giúp KNN tính toán khoảng cách chính xác hơn và đạt được độ chính xác cao.
- 2. Đối với bài toán nhận dạng ký tự (Letter Recognition):

Độ chính xác (Accuracy): 0.9425 (khoảng 94.25%).
Nhận xét: Đạt được độ chính xác trên 94% trên tập dữ liệu gồm 20,000 mẫu và 26 lớp ký tự là một thành công đáng kể cho thuật toán KNN. Điều này cho thấy KNN có khả năng xử lý tốt các bài toán phân loại đa lớp với số lượng mẫu lớn và nhiều đặc trưng (16 đặc trưng). Tương tự như Iris, việc chuẩn hóa dữ liệu là yếu tố then chốt để mô hình hoạt động hiệu quả trên tập dữ liệu này, nơi các đặc trưng có thể có các thang đo khác nhau.
Tổng kết và Điểm mạnh của KNN trong các bài toán này:

Đơn giản và dễ hiểu: KNN là một thuật toán rất trực quan, không yêu cầu giả định phức tạp về phân phối dữ liệu.
Hiệu quả cao: Cả hai thí nghiệm đều cho thấy KNN đạt được độ chính xác ấn tượng, đặc biệt là với Letter Recognition, chứng tỏ tính mạnh mẽ của nó ngay cả trên các tập dữ liệu phức tạp hơn.
Quan trọng của chuẩn hóa dữ liệu: Kết quả từ cả hai bài toán đều nhấn mạnh tầm quan trọng của việc chuẩn hóa dữ liệu. KNN dựa trên khoảng cách, nên các đặc trưng cần có cùng một thang đo để tránh việc các đặc trưng có giá trị lớn hơn chi phối phép tính khoảng cách.
Hạn chế (được ngụ ý):

Chi phí tính toán: Mặc dù không thể hiện rõ trong báo cáo, nhưng KNN có thể chậm khi dự đoán với tập dữ liệu rất lớn, vì nó phải tính khoảng cách tới tất cả các điểm huấn luyện. Tuy nhiên, với 20,000 mẫu của Letter Recognition, tốc độ vẫn chấp nhận được.
Nhạy cảm với nhiễu và outlier: Giá trị của k (số lượng láng giềng) rất quan trọng. k nhỏ có thể khiến mô hình nhạy cảm với nhiễu, trong khi k lớn có thể làm mất đi các chi tiết quan trọng. Trong các thí nghiệm này, k=5 dường như là một lựa chọn tốt.


### Câu 2. Nhận dạng ký tự dùng thuật toán K Nearest Neighbors.
The Database: UCI Letter-Recognition Data in a text file.
26 classes: A to Z
16-D feature vectors
20,000 samples
Tham khảo: https://archive.ics.uci.edu/ml/datasets/letter+recognition


## load dữ liệu


In [ ]:
import pandas as pd

url = "Data_letter/letter-recognition.data"
cols = ["letter","x-box","y-box","width","high","onpix","x-bar","y-bar",
        "x2bar","y2bar","xybar","x2ybr","xy2br","x-ege","xegvy","y-ege","yegvx"]

df = pd.read_csv(url, names=cols)

## Tách dữ liệu

In [ ]:
from sklearn.model_selection import train_test_split

X = df.drop("letter", axis=1)
y = df["letter"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

## Chuẩn hóa dữ liệu


In [ ]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

## Huấn luyện mô hình KNN

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

model = KNeighborsClassifier(n_neighbors=5)  # k = 5 là phổ biến
model.fit(X_train, y_train)

KNeighborsClassifier()

## Đánh giá mô hình KNN

In [ ]:
from sklearn.metrics import accuracy_score

y_pred = model.predict(X_test)
acc = accuracy_score(y_test, y_pred)
print("Accuracy:", acc)

Accuracy: 0.9425


## Thử nghiệm mô hình trên dữ liệu mẫu

In [ ]:
import numpy as np

sample = np.array([[2,8,3,5,1,8,6,4,3,2,1,7,6,5,8,3]])
sample = scaler.transform(sample)
print(model.predict(sample))

['A']


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


📘 BÁO CÁO BÀI TẬP: Nhận dạng ký tự dùng thuật toán K-Nearest Neighbors (KNN)
1. Giới thiệu bài toán

Bài toán yêu cầu xây dựng mô hình nhận dạng ký tự (A–Z) dựa trên thuật toán K-Nearest Neighbors.
Dữ liệu sử dụng là UCI Letter Recognition Data Set, bao gồm 20.000 mẫu đã được trích xuất đặc trưng từ ảnh chữ in.

2. Mô tả dữ liệu

Số mẫu: 20.000

Số lớp: 26 ký tự từ A → Z

Số đặc trưng: 16 thuộc tính dạng số nguyên (0–15)

Định dạng mỗi dòng:
class, x-box, y-box, width, high, onpix, x-bar, y-bar,
x2bar, y2bar, xybar, x2ybr, xy2br, x-ege, xegvy, y-ege, yegvx

Biến mục tiêu (label): ký tự chữ cái A–Z

Dữ liệu gốc: đã được trích xuất đặc trưng nên không cần tiền xử lý ảnh.

3. Thuật toán K-Nearest Neighbors

KNN là thuật toán phân loại dựa trên khoảng cách giữa điểm cần dự đoán và các điểm trong tập huấn luyện.

Ý tưởng:

Tính khoảng cách từ mẫu mới đến tất cả các mẫu đã biết.

Chọn k mẫu gần nhất.

Dự đoán nhãn theo đa số phiếu.

KNN không huấn luyện theo nghĩa truyền thống ➝ chi phí tính toán nằm ở bước dự đoán.

KNN nhạy cảm với đơn vị đo, do đó cần chuẩn hoá dữ liệu.

4. Các bước thực hiện
4.1. Tải và đọc dữ liệu

Dữ liệu được đọc từ file letter-recognition.data với 17 cột (1 label + 16 features).

4.2. Tách train/test

Dữ liệu được chia:

80% dùng để huấn luyện

20% dùng để kiểm tra mô hình

4.3. Chuẩn hoá dữ liệu

Sử dụng StandardScaler

Chuyển dữ liệu về phân phối có trung bình = 0, độ lệch chuẩn = 1

Việc chuẩn hoá giúp KNN đo khoảng cách tốt hơn.

4.4. Huấn luyện mô hình

Sử dụng:

KNeighborsClassifier(n_neighbors=5)


k=5 là lựa chọn phổ biến, tránh overfitting.

4.5. Dự đoán và đánh giá

Dùng accuracy để đánh giá mức độ chính xác của mô hình.

Mô hình KNN trên tập dữ liệu này đạt:

Accuracy 0.9425

5. Kết quả nhận xét

Mô hình KNN đạt độ chính xác cao, phù hợp với bài toán nhận dạng ký tự dựa trên đặc trưng số.

Độ chính xác tốt ngay cả khi cấu hình đơn giản (k=5).

Điểm mạnh:

Dễ cài đặt

Không cần quá trình huấn luyện phức tạp

Điểm yếu:

Tốc độ dự đoán chậm khi dữ liệu lớn (20.000 mẫu → vẫn chấp nhận được)

Nhạy cảm với scale của dữ liệu → bắt buộc phải chuẩn hoá

6. Kết luận

Thuật toán KNN hoạt động hiệu quả trên bộ dữ liệu UCI Letter Recognition.

Với các đặc trưng đã được chuẩn hoá, mô hình KNN có thể phân loại 26 chữ cái A–Z với độ chính xác trên 92%.

KNN là một baseline mạnh cho các bài toán nhận dạng ký tự dạng bảng.

## THỰC HÀNH VỀ PHÂN LỚP BAYES


### Câu 1. Toy example: Phân loại loài hoa dựa trên tập dữ liệu Iris. Cài đặt chương trình demo thuật toánNaïve Bayes. Tham khảo: https://alphacoder.xyz/naive-bayes, https://archive.ics.uci.edu/ml/datasets/iris

In [ ]:
import numpy as np
from sklearn import datasets

# 1. CHUẨN BỊ DỮ LIỆU (Thay vì nhập tay, ta load từ thư viện cho nhanh)
# Load dataset Iris
iris = datasets.load_iris()
X = iris.data  # Các đặc trưng (4 cột: sepal length, sepal width, petal length, petal width)
y = iris.target # Nhãn lớp (0: Setosa, 1: Versicolor, 2: Virginica)
class_names = iris.target_names # Tên các loài hoa

print("Dữ liệu mẫu (hàng đầu tiên):", X[0])
print("Nhãn mẫu:", class_names[y[0]])
print("-" * 30)

# 2. TÍNH P(CLASS) - Xác suất tiên nghiệm
# Đếm số lượng mẫu của từng lớp
sample_size = len(y)
class_counts = {} # Lưu số lượng mỗi lớp

# Khởi tạo đếm
for c in np.unique(y):
    class_counts[c] = 0

for label in y:
    class_counts[label] += 1

# Tính xác suất P(Class)
p_classes = {}
for c in class_counts:
    p_classes[c] = class_counts[c] / sample_size

print("P(Class):", p_classes)

# 3. TÍNH TRUNG BÌNH (MEAN) VÀ PHƯƠNG SAI (VARIANCE) CHO TỪNG ĐẶC TRƯNG CỦA TỪNG LỚP
# Logic tương tự đoạn: mean_height_males = np.mean(heights_of_males)...
means = {}
vars = {}

# Tách dữ liệu theo từng lớp để tính toán
for c in np.unique(y):
    # Lấy tất cả các hàng thuộc lớp c
    X_c = X[y == c]

    # Tính mean và var cho từng cột (từng đặc trưng)
    # axis=0 nghĩa là tính dọc theo cột
    means[c] = np.mean(X_c, axis=0)
    vars[c] = np.var(X_c, axis=0)

# 4. HÀM TÍNH PDF (GIỐNG HỆT BÀI MẪU)
def the_pdf(x, mean, variance):
    # Công thức phân phối chuẩn Gaussian
    pd = 1 / (np.sqrt(2 * np.pi * variance)) * np.exp((-(x - mean)**2) / (2 * variance))
    return pd

# 5. DỰ ĐOÁN (PREDICT)
# Giả sử ta có một bông hoa mới với thông số này:
# [Sepal Length, Sepal Width, Petal Length, Petal Width]
input_flower = [5.8, 3.0, 5.1, 1.8]

print(f"\nĐang dự đoán cho mẫu: {input_flower}")

best_class = None
max_prob = -1

# Duyệt qua từng lớp (0, 1, 2) để tính xác suất
for c in np.unique(y):
    # Lấy P(Class)
    prior = p_classes[c]

    # Tính P(Feature | Class) cho từng đặc trưng bằng hàm PDF
    # Vì Iris có 4 đặc trưng, ta phải nhân 4 lần PDF (tương tự bài mẫu nhân height * weight)
    p_feature_0 = the_pdf(input_flower[0], means[c][0], vars[c][0])
    p_feature_1 = the_pdf(input_flower[1], means[c][1], vars[c][1])
    p_feature_2 = the_pdf(input_flower[2], means[c][2], vars[c][2])
    p_feature_3 = the_pdf(input_flower[3], means[c][3], vars[c][3])

    # Công thức Naive Bayes: P(Class|Data) = P(Class) * P(f1|C) * P(f2|C)...
    total_probability = prior * p_feature_0 * p_feature_1 * p_feature_2 * p_feature_3

    print(f"Xác suất là {class_names[c]}: {total_probability}")

    # Tìm lớp có xác suất cao nhất
    if total_probability > max_prob:
        max_prob = total_probability
        best_class = c

print("-" * 30)
print(f"KẾT QUẢ: Bông hoa này được phân loại là: {class_names[best_class]}")

Dữ liệu mẫu (hàng đầu tiên): [5.1 3.5 1.4 0.2]
Nhãn mẫu: setosa
------------------------------
P(Class): {np.int64(0): 0.3333333333333333, np.int64(1): 0.3333333333333333, np.int64(2): 0.3333333333333333}

Đang dự đoán cho mẫu: [5.8, 3.0, 5.1, 1.8]
Xác suất là setosa: 5.386151412713544e-147
Xác suất là versicolor: 0.00447707314828265
Xác suất là virginica: 0.06473767235250426
------------------------------
KẾT QUẢ: Bông hoa này được phân loại là: virginica


#### Báo cáo: Phân loại hoa Iris sử dụng Gaussian Naive Bayes

## 1. Mục tiêu và Phương pháp
* **Mục tiêu:** Phân loại 3 loài hoa (Setosa, Versicolor, Virginica) dựa trên 4 đặc trưng: Chiều dài/rộng đài hoa và Chiều dài/rộng cánh hoa.
* **Phương pháp:** Tự cài đặt thuật toán **Gaussian Naive Bayes** từ đầu (không dùng `sklearn.naive_bayes` để train) nhằm hiểu rõ bản chất toán học.
* **Lý do chọn Gaussian:** Vì các đặc trưng của hoa Iris là dữ liệu liên tục (số thực), ta giả định chúng tuân theo phân phối chuẩn (Normal Distribution).

## 2. Tóm tắt quy trình cài đặt
1.  **Tính tham số thống kê:** Tính trung bình ($\mu$) và phương sai ($\sigma^2$) cho từng đặc trưng của mỗi lớp hoa từ tập dữ liệu huấn luyện.
2.  **Tính xác suất tiên nghiệm (Prior):** $P(Class)$ dựa trên tỷ lệ số lượng mẫu của từng lớp.
3.  **Hàm mật độ xác suất (PDF):** Sử dụng công thức Gaussian PDF để tính xác suất của một đặc trưng rơi vào một lớp:
    $$P(x|c) = \frac{1}{\sqrt{2\pi\sigma^2}} \cdot e^{-\frac{(x-\mu)^2}{2\sigma^2}}$$
4.  **Dự đoán:** Với một mẫu hoa mới, tính xác suất tổng hợp cho từng lớp và chọn lớp có xác suất cao nhất (Maximum A Posteriori).

## 3. Nhận xét và Kết luận
* **Kết quả:** Thuật toán tính ra được xác suất cụ thể cho từng lớp. Lớp nào có xác suất cao vượt trội sẽ là nhãn dự đoán.
* **Ưu điểm:**
    * Cài đặt đơn giản, dễ hiểu.
    * Tốc độ tính toán cực nhanh (chỉ cần tính mean và variance một lần).
    * Hiệu quả tốt ngay cả với tập dữ liệu nhỏ như Iris.
* **Hạn chế:** Thuật toán được gọi là "Naive" (Ngây thơ) vì nó giả định các đặc trưng **độc lập** với nhau (ví dụ: coi chiều dài cánh và chiều rộng cánh không liên quan). Thực tế các thông số này thường tương quan, nhưng thuật toán vẫn cho kết quả khá chính xác.

### Câu 2. Nhận dạng ký tự dùng thuật toán Naïve Bayes.
The Database: UCI Letter-Recognition Data in a text file.
26 classes: A to Z
16-D feature vectors
20,000 samples
Tham khảo: https://archive.ics.uci.edu/ml/datasets/letter+recognition

In [ ]:
#code here
import numpy as np
import pandas as pd

class Naive_Bayes:
    def __init__(self, data_set):
        self.ds = data_set
        self.features = self.ds.columns[1:]  # 16 features

        # tính mean và variance theo từng lớp
        self.ds_means = self.ds.groupby('letter').mean()
        self.ds_variances = self.ds.groupby('letter').var()
        self.class_probabilities = self.get_class_probabilities(self.ds)

    def get_class_probabilities(self, data_set):
        class_sizes = data_set.groupby('letter').size()
        total = len(data_set)
        return {cls: count / total for cls, count in class_sizes.items()}

    def get_probability_density(self, x, mean, variance):
        variance = max(variance, 1e-6)  # tránh chia 0
        exponent = -((x - mean) ** 2) / (2 * variance)
        return (1 / np.sqrt(2 * np.pi * variance)) * np.exp(exponent)

    def predict(self, x):
        log_probs = {}

        for cls, prior in self.class_probabilities.items():
            log_prob = np.log(prior)

            for i, feat in enumerate(self.features):
                mean = self.ds_means.loc[cls, feat]
                var = self.ds_variances.loc[cls, feat]
                pdf = self.get_probability_density(x[i], mean, var)
                log_prob += np.log(pdf)

            log_probs[cls] = log_prob

        return max(log_probs, key=log_probs.get)

    def test(self, test_data):
        correct = 0

        for row in test_data.itertuples(index=False):
            true_label = row[0]
            x = row[1:]
            pred = self.predict(x)
            if pred == true_label:
                correct += 1

        acc = correct / len(test_data)
        print("Accuracy =", acc)


In [ ]:
# ====== 1. TÊN CỘT CHUẨN THEO UCI ======
col_names = [
    'label', 'f1', 'f2', 'f3', 'f4', 'f5', 'f6',
    'f7', 'f8', 'f9', 'f10', 'f11', 'f12',
    'f13', 'f14', 'f15', 'f16'
]

# ====== 2. TRỘN NGẪU NHIÊN ======
lc_datga = lc_datga.sample(frac=1, random_state=42).reset_index(drop=True)

# ====== 3. CHIA TRAIN : TEST = 80 : 20 ======
split = int(0.8 * len(lc_datga))
train_data = lc_datga.iloc[:split].reset_index(drop=True)
test_data = lc_datga.iloc[split:].reset_index(drop=True)

# ====== 4. HUẤN LUYỆN NAÏVE BAYES ======
nb = Naive_Bayes(train_data)

# ====== 5. ĐÁNH GIÁ ======
nb.test(test_data)


Accuracy = 0.64175


# Báo cáo tóm tắt: Nhận dạng ký tự với Naïve Bayes

## 1. Tổng quan bài toán
* **Mục tiêu:** Phân loại 26 chữ cái (A-Z) dựa trên dataset UCI Letter Recognition.
* **Dữ liệu:** 20,000 mẫu, mỗi mẫu có 16 đặc trưng số (feature) liên tục.
* **Phương pháp:** Gaussian Naïve Bayes (do đặc trưng là số thực).

## 2. Kỹ thuật cài đặt (Code Highlights)
* **Huấn luyện (Training):** Tính toán Mean ($\mu$) và Variance ($\sigma^2$) cho từng đặc trưng của mỗi lớp bằng `groupby` (Pandas) để tận dụng vector hóa, tăng tốc độ xử lý.
* **Dự đoán (Prediction):** Sử dụng tổng Log-Likelihood thay vì nhân xác suất trực tiếp để tránh lỗi *underflow* (tràn số âm).
* **Xử lý lỗi:** Thêm một giá trị nhỏ `epsilon = 1e-6` vào phương sai để tránh lỗi chia cho 0.

## 3. Nhận xét
* **Ưu điểm:** Thuật toán đơn giản, chi phí tính toán thấp, tốc độ huấn luyện và dự đoán cực nhanh.
* **Nhược điểm:** Giả định các đặc trưng độc lập (Independence Assumption) không hoàn toàn đúng với dữ liệu hình ảnh ký tự, có thể giới hạn độ chính xác so với các mô hình phức tạp hơn (SVM, Neural Networks).
* **Kết luận:** Đây là mô hình baseline hiệu quả để đánh giá sơ bộ dữ liệu trước khi áp dụng các thuật toán phức tạp.